rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,438 @@
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol so would be
+internalized to allow for changes (so long as this wouldn't change the overall
+behavior).
+
+### Current Use of Versioning
+
+Records of versions are currently stored in `State` and in the `Header` of each
+block. This already covers a broad base of scenarios. The consensus reactor can
+use the version from state to ensure it reads `Vote`s and `Partsetheader`s
+correctly. Similarly, fast sync, the block executor and evidence module, whom
+all interact with block protocol data structures also all have access to state.
+The next few sections will explore the aspects of Tendermint that will need to
+be modified to prepare for multi versioning support.
+
+### RPC
+
+Supporting soft upgrades means that the RPC will need to be able to deliver
+information across different versions. The RPC will therefore need to know which
+heights corresponds to which version and then be able to communicate this to
+the clients. This would most likely require clients to be versatile to handling
+the different block versions. For example block explorers would also take on the
+burden of supporting all block versions across a chain.
+
+### Light Clients
+
+Similar to RPC, light clients will also need to be wary of block protocol
+changes and be able to handle them accordingly. Depending on these changes,
+there might be the need for special verification logic across signed headers
+with different versions.
+
+### Implications
+
+**Positive:**
+
+- It requires no changes to the database therefore keeps the immutable aspect
+  of the blockchain.
+- Async upgrading. Nodes choose when they want to upgrade the binary and
+Tendermint handles the actual transition.
+
+**Negative:**
+
+- Increased burden on RPC clients to be able to parse multiple versions
+- Bloating of code base and binary. We would also need to figure out methods
+of maintaining good code hygiene with respect to having all these versions.
+- If we separate versions i.e. ABCI and Block and P2P this could potentially
+compound the problem if we need to consider the different permutations and
+how they can be supported.
```

Comment:
Correctness analysis and proper testing need to consider cross-version scenarios. With several supported versions, this can be very complex and error prone task.